File: Coding/Ruby/sum_types.rb
# Different methods of using loops to sum an array of numbers

def recursion(array, sum)
    if array.empty?
        return sum
    else
        sum += array.first
        return recursion(array[1..-1], sum)
    end
end

def for_loop(array)
    sum = 0
    
    for i in array
        sum += i
    end
    
    return sum
end

def while_loop(array)
    sum = 0
    i = 0
    
    while i<array.length
        sum += array[i]
        i += 1
    end
    
    return sum
end

elements = [1,2,3,4,5,6,7,8,9,10]
print "#{recursion(elements, 0)}\n"
print "#{for_loop(elements)}\n"
print "#{while_loop(elements)}\n"

File: Coding/C++/sum_types.cpp
#include<iostream>

using namespace std;

int recursive(int [], int, int);
int for_loop(int [],int);
int while_loop(int [],int);

int main(){
	int elements[] = {1,2,3,4,5,6,7,8,9,10};//{5,15,23,68,91,12,35,46,81,3,12345,100,513,214,79,651,5,17,54,89};
	int n = sizeof(elements)/sizeof(int)-1;
	int sum = 0;
	sum = recursive(elements, n, sum);
	cout << sum << " is the recursive answer.\n";
	sum = for_loop(elements, n);
	cout << sum << " is the for loop answer.\n";
	sum = while_loop(elements, n);
	cout << sum << " is the while loop answer.\n";
	return 0;
}

int recursive(int param[], int size, int sum){
	if(size>=0){
		sum += param[size];
	} else{
		return sum;
	}
	size--;
	return recursive(param, size, sum);
}

int for_loop(int param[], int size){
	int i, sum = 0;
	
	for(i=0;i<=size;i++){
		sum += param[i];
	}
	return sum;
}

int while_loop(int param[],int size){
	int i = 0, sum = 0;

	while(i <= size){
		sum+=param[i];
		i++;
	}
	return sum;
}